controllers/openstackclient_controller.go

```diff
@@ -441,122 +496,85 @@ func (r *OpenStackClientReconciler) podCreateOrUpdate(
-			},
-		}
-		initEnv = append(initEnv, idmEnv...)
-	}
-
-	initContainerDetails := []openstackclient.InitContainer{
+	pod.Spec.InitContainers = []corev1.Container{
 		{
-			ContainerImage: instance.Spec.ImageURL,
-			Env:            initEnv,
-			Privileged:     false,
-			VolumeMounts:   initVolumeMounts,
+			Name: "init-0",
 		},
 	}
 
-	pod.Spec.InitContainers = openstackclient.GetInitContainers(initContainerDetails)
-
 	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, pod, func() error {
-		pod.Spec.Containers[0].Env = common.MergeEnvs(pod.Spec.Containers[0].Env, *envVars)
+		// Note: initialize the pod structs above then only reconcile individual fields here
+		// Can not add/replace new structs in the pod here as that will drop the defaults that
+		// are set/added when the pod is created.
+
+		isPodUpdate := !pod.ObjectMeta.CreationTimestamp.IsZero()
 
-		// labels
-		common.InitMap(&pod.Labels)
+		pod.ObjectMeta.Annotations["k8s.v1.cni.cncf.io/networks"] = annotation
 		for k, v := range common.GetLabels(instance, openstackclient.AppLabel, map[string]string{}) {
 			pod.Labels[k] = v
 		}
 
+		if isPodUpdate {
+			// must merge envs/volumes etc... when the pod already exists
+			env = common.MergeEnvs(pod.Spec.Containers[0].Env, *envVars)
+			initEnv = common.MergeEnvs(pod.Spec.InitContainers[0].Env, initEnvVars)
+			volumeMounts = common.MergeVolumeMounts(pod.Spec.Containers[0].VolumeMounts, volumeMounts)
+			initVolumeMounts = common.MergeVolumeMounts(pod.Spec.InitContainers[0].VolumeMounts, initVolumeMounts)
+			volumes = common.MergeVolumes(pod.Spec.Volumes, volumes)
+		}
+
+		pod.Spec.SecurityContext.RunAsUser = &runAsUser
+		pod.Spec.SecurityContext.RunAsGroup = &runAsGroup
+		pod.Spec.SecurityContext.FSGroup = &runAsGroup
+		pod.Spec.ServiceAccountName = openstackclient.ServiceAccount
+		pod.Spec.TerminationGracePeriodSeconds = &terminationGracePeriodSeconds
+		pod.Spec.Volumes = volumes
+		pod.Spec.Containers[0].Image = instance.Spec.ImageURL
+		pod.Spec.Containers[0].ImagePullPolicy = corev1.PullAlways
+		pod.Spec.Containers[0].Env = env
+		pod.Spec.Containers[0].VolumeMounts = volumeMounts
+		pod.Spec.InitContainers[0].Image = instance.Spec.ImageURL
+		pod.Spec.InitContainers[0].ImagePullPolicy = corev1.PullAlways
+		pod.Spec.InitContainers[0].Env = initEnv
+		pod.Spec.InitContainers[0].VolumeMounts = initVolumeMounts
+
+		if len(instance.Spec.DNSServers) != 0 {
+			pod.Spec.DNSPolicy = corev1.DNSNone
+			pod.Spec.DNSConfig.Nameservers = instance.Spec.DNSServers
+		} else {
+			pod.Spec.DNSPolicy = corev1.DNSClusterFirst
+			pod.Spec.DNSConfig.Nameservers = []string{}
+		}
+		if len(instance.Spec.DNSSearchDomains) != 0 {
+			pod.Spec.DNSConfig.Searches = instance.Spec.DNSSearchDomains
+		} else {
+			pod.Spec.DNSConfig.Searches = []string{}
+		}
+
+		if isPodUpdate && len(pod.Status.InitContainerStatuses) > 0 {
+			if pod.Status.InitContainerStatuses[0].Image != pod.Spec.InitContainers[0].Image {
```

Comment:
just to mention another possibility, could also check this in the webhook where we have the old + new spec to see if it changed and set some status field.